controllers/idler/aap_idler.go

```diff
@@ -0,0 +1,252 @@
+		return false
+	}
+	for _, aap := range allAAPs.Items {
+		found := false
+		for _, idledAAP := range idledAAPs {
+			if aap.GetName() == idledAAP {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return false
+		}
+	}
+	return true
+}
+
+const oneHour = 60 * 60 // in seconds
+
+func (r *Reconciler) aapTimeoutSeconds(idler *toolchainv1alpha1.Idler) int32 {
+	// Check if the idler timeout is less than one hour and if so, set it to half of the timeout.
+	// Otherwise, subtract one hour from the timeout.
+	// This is to ensure that the AAP idler kicks in before the main idler.
+	timeoutSeconds := idler.Spec.TimeoutSeconds
+	if timeoutSeconds <= oneHour {
+		timeoutSeconds = timeoutSeconds / 2
+	} else {
+		timeoutSeconds = timeoutSeconds - oneHour
+	}
+	return timeoutSeconds
+}
+
+// ensureAAPIdled checks if the long-running or crash-looping pod belongs to an AAP instance and if so, ensures that the AAP is idled.
+// Returns the AAP resource name in case it was idled, or an empty string if it was not idled.
+func (r *Reconciler) ensureAAPIdled(ctx context.Context, pod corev1.Pod, aapAPI aapAPI) (string, error) {
+	podCondition := pod.Status.Conditions
+	for _, podCond := range podCondition {
+		if podCond.Type == "PodCompleted" {
+			// Pod is in the completed state, no need to idle
+			return "", nil
+		}
+	}
+
+	aap, err := r.getAAPOwner(ctx, &pod, aapAPI)
+	if err != nil {
+		return "", err
+	}
+	if aap == nil {
+		return "", nil // No AAP owner found
+	}
+
+	// Patch the aap resource by setting spec.idle_aap to true in order to idle it
+	patch := []byte(`{"spec":{"idle_aap":true}}`)
+	_, err = r.DynamicClient.Resource(aapAPI.GVR).Namespace(pod.Namespace).Patch(ctx, aap.GetName(), types.MergePatchType, patch, metav1.PatchOptions{})
+	if err != nil {
+		return "", err
+	}
+
+	log.FromContext(ctx).Info("AAP idled", "name", aap.GetName())
+
+	return aap.GetName(), nil
+}
+
+// getAAPOwner returns the top level owner of the given object if it is an AAP instance.
+func (r *Reconciler) getAAPOwner(ctx context.Context, obj metav1.Object, aapAPI aapAPI) (metav1.Object, error) {
+	owners := obj.GetOwnerReferences()
+	var owner metav1.OwnerReference
+	if len(owners) == 0 {
+		return nil, nil // No owner
+	}
+	if len(owners) > 1 {
+		owner = owners[0] // Multiple owners, use the first one
+	}
+	logger := log.FromContext(ctx)
+	// Get the GVR for the owner
+	gvr, err := gvrForKind(owner.Kind, owner.APIVersion, aapAPI)
+	if err != nil {
+		return nil, err
+	}
+	// Get the owner object
+	ownerObject, err := r.DynamicClient.Resource(*gvr).Namespace(obj.GetNamespace()).Get(ctx, owner.Name, metav1.GetOptions{})
+	if errors.IsNotFound(err) { // Ignore not found errors. Can happen if the parent controller has been deleted. The Garbage Collector should delete the pods shortly.
+		logger.Info("Owner not found", "kind", owner.Kind, "name", owner.Name)
+		return nil, nil
+	}
+	if ownerObject.GetKind() == "AnsibleAutomationPlatform" {
```

Comment:
we could  reuse  the constant here
```suggestion
	if ownerObject.GetKind() == aapKind {
```